UsageStatistics_StorageWriter.php

```diff
@@ -1,368 +1,504 @@
+		$metric_values                    = array();
 		$metric_values['timestamp_start'] = get_site_option( 'w3tc_stats_hotspot_start' );
-		$metric_values['timestamp_end'] = $hotspot_endtime_int;
+		$metric_values['timestamp_end']   = $hotspot_endtime_int;
 
-		// try to limit time between get and reset of counter value
-		// to loose as small as posssible
+		// try to limit time between get and reset of counter value to loose as small as posssible.
 		foreach ( $metrics as $metric ) {
-			$metric_values[$metric] = $this->cache_storage->counter_get( $metric );
+			$metric_values[ $metric ] = $this->cache_storage->counter_get( $metric );
 			$this->cache_storage->counter_set( $metric, 0 );
 		}
 
-		$metric_values = apply_filters( 'w3tc_usage_statistics_metric_values',
-			$metric_values );
+		$metric_values = apply_filters( 'w3tc_usage_statistics_metric_values', $metric_values );
 
 		$history_encoded = get_site_option( 'w3tc_stats_history' );
-		$history = null;
-		if ( !empty( $history_encoded ) )
+		$history         = null;
+		if ( ! empty( $history_encoded ) ) {
 			$history = json_decode( $history_encoded, true );
-		if ( !is_array( $history ) )
+		}
+
+		if ( ! is_array( $history ) ) {
 			$history = array();
+		}
 
 		$time_keep_border = time() - $this->keep_history_interval_seconds;
 
-		if ( $hotspot_endtime_int < $time_keep_border )
+		if ( $hotspot_endtime_int < $time_keep_border ) {
 			$history = array(
 				array(
 					'timestamp_start' => $time_keep_border,
-					'timestamp_end' => (int)$this->new_hotspot_endtime -
-					$this->slot_interval_seconds - 1
-				)
-			);   // this was started too much time from now
-		else {
-			// add collected
+					'timestamp_end'   => (int) $this->new_hotspot_endtime - $this->slot_interval_seconds - 1,
+				),
+			); // this was started too much time from now.
+		} else {
+			// add collected.
 			$history[] = $metric_values;
 
-			// if we empty place later - fill it
+			// if we empty place later - fill it.
 			for ( ;; ) {
-				$metric_values = array(
-					'timestamp_start' => $metric_values['timestamp_end']
+				$metric_values                  = array(
+					'timestamp_start' => $metric_values['timestamp_end'],
 				);
-				$metric_values['timestamp_end'] =
-					$metric_values['timestamp_start'] + $this->slot_interval_seconds;
-				if ( $metric_values['timestamp_end'] < $this->now )
+				$metric_values['timestamp_end'] = $metric_values['timestamp_start'] + $this->slot_interval_seconds;
+				if ( $metric_values['timestamp_end'] < $this->now ) {
 					$history[] = $metric_values;
-				else
+				} else {
 					break;
+				}
 			}
 
-			// make sure we have at least one value in history
-			for ( ;count( $history ) > $this->slots_count; ) {
-				if ( !isset( $history[0]['timestamp_end'] ) ||
-					$history[0]['timestamp_end'] < $time_keep_border )
+			// make sure we have at least one value in history.
+			$history_count = count( $history );
+			while ( $history_count > $this->slots_count ) {
```

Comment:
This will cause an indefinite loop.  The `while` condition must check the `count` of `$history` for every iteration.  You can change it to `while ( true )` and `break` if the condition is met.

Reply:
Incorrect. It shifts the array and updates the count. Shouldn't be infinite

`// make sure we have at least one value in history.
			$history_count = count( $history );
			while ( $history_count > $this->slots_count ) {
				if ( ! isset( $history[0]['timestamp_end'] ) || $history[0]['timestamp_end'] < $time_keep_border ) {
					array_shift( $history );
				} else {
					break;
				}

				// Update the history count after modification.
				$history_count = count( $history );
			}`